### PR TITLE
JDK-8298448: UndefinedBehaviorSanitizer

### PR DESCRIPTION
**Pull request by jcking**
Allow building OpenJDK with UBSan. Currently the build fails when optimizing the image due to lots of undefined behavior (it invokes the built JVM). Follow up PRs will either replace the undefined behavior with well defined behavior or suppress errors which are intentional. The goal is to make OpenJDK more well defined and thus more portable across compilers and architectures.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8298448](https://bugs.openjdk.org/browse/JDK-8298448): UndefinedBehaviorSanitizer

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [d878ede0](https://git.openjdk.org/jdk/pull/11604/files/d878ede08794f3e304582788093e80f3187e531e)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11604/head:pull/11604` \
`$ git checkout pull/11604`

Update a local copy of the PR: \
`$ git checkout pull/11604` \
`$ git pull https://git.openjdk.org/jdk pull/11604/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11604`

View PR using the GUI difftool: \
`$ git pr show -t 11604`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11604.diff">https://git.openjdk.org/jdk/pull/11604.diff</a>

</details>

### CODE REVIEW

**Comment by robehn** [comment]
Two questions, this works fine for me:
`UBSAN_CFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero -Wno-stringop-truncation -Wno-format-overflow -fno-omit-frame-pointer" ; UBSAN_LDFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero" ; sh configure --with-conf-name=udf --with-debug-level=fastdebug --with-native-debug-symbols=internal -with-version-opt=releaserobbin --with-gtest=/home/rehn/source/gtest181/ --disable-precompiled-headers --enable-dtrace --with-extra-cflags="$UBSAN_CFLAGS" --with-extra-cxxflags="$UBSAN_CFLAGS" --with-extra-ldflags="$UBSAN_LDFLAGS" && make images CONF=udf`
I.e.build works ?

Secondly, the method __ubsan_default_options is in the java launcher.
libjvm.so is a "standalone product" that can be used by different launchers.
So I'm not sure it should be in launcher if it was needed.

**Comment by jcking** [comment]
What version of GCC are you using? I'm on 12.2 which is the most recent. It
might be that I'm on a newer build with more UBSan coverage.

__ubsan_default_options is called extremely early before main, during
static initializers. IIRC the launcher uses dynamic loading to get
libjvm.so, so that won't work. But maybe I'm misremembering, I'll double
check.

**Comment by robehn** [comment]
> What version of GCC are you using?

gcc 11.3 with libubsan 11.2

Also it seem to big overlap with -Wcast-align(=strict) for the warnings/errors I see and I do like that warning.
Do you have idea if the coverage are pretty much the same this particular warning?
I.e. is anyone of them better than the other?

File: make/autoconf/jdk-options.m4
Comment (erikj79) [inline make/autoconf/jdk-options.m4:450]: I think this logic fits better in `flags.m4`, otherwise this looks ok to me.

**Comment by theRealAph** [comment]
> > What version of GCC are you using?
> 
> gcc 11.3 with libubsan 11.2
> 
> Also it seem to big overlap with -Wcast-align(=strict) for the warnings/errors I see and I do like that warning. Do you have an idea if the coverage are pretty much the same for this particular warning? I.e. is anyone of them better than the other?

You can't really compare them because Undefined Behaviour is a runtime thing rather than a compile-time thing. Something like `x = *(char*)0;` is legal in a conforming C++ program as long as it's never executed.

I vastly prefer UB Sanitizer because it detects a ton of stuff that static analysis can't. But there's plenty of room for both.

**Comment by theRealAph** [comment]
I'm really pleased to see this.

Im working on `https://github.com/openjdk/jdk/pull/10920`, which fixes the last of the null pointer dereferences in x86/AArch64 HotSpot. Once that's done I'm thinking of enabling `-fsanitize=null` by default in debug builds, as long as it's not too slow.

File: make/autoconf/jdk-options.m4
Comment (erikj79) [reply make/autoconf/jdk-options.m4:450]: Ah now I understand that this compiles runtime checks into the product. In that case it does actually fit well into jdk-options.m4, so you can leave it there.

**Review by erikj79 — APPROVED** [review]

File: make/autoconf/spec.gmk.in
Comment (dholmes-ora) [inline make/autoconf/spec.gmk.in:459]: I don't see anything reading this. ??

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [inline src/java.base/share/native/launcher/main.c:37]: I really do not like having to make source code changes to accommodate these kinds of tools.

**Review by dholmes-ora — COMMENTED** [review]
I think it requires much broader discussion as to whether OpenJDK is actively seen to endorse these tools. Why these tools? What if there are other tools, should we support them all?

I'm not saying use of these tools may not be useful, but actually incorporating them into OpenJDK is a decision that needs to be made at a higher-level IMO.

**Comment by jcking** [comment]
> I think it requires much broader discussion as to whether OpenJDK is actively seen to endorse these tools. Why these tools? What if there are other tools, should we support them all?
> 
> I'm not saying use of these tools may not be useful, but actually incorporating them into OpenJDK is a decision that needs to be made at a higher-level IMO.

The sanitizers are integrated directly with GCC and Clang/LLVM and are used by projects such as the Linux kernel. They are also used by companies such as Facebook and Google, which IIRC maintain some of the largest closed source mono repositories on the planet. As the sanitizers are integrated directly with GCC and Clang/LLVM, they are extremely easy to use (no external dependencies), fast, and have no direct alternatives. An alternative would also need to be integrated with the compilers in order to be at par.

Additionally configuration options for using ASan already exist in OpenJDK, so that ship has kinda sailed.

If we feel strongly about a discussion, we should probably discuss all the sanitizers as a whole. However that discussion can be done in parallel, as ASan is already used. Just adding the options to OpenJDK does not mean it is endorsed.

**Comment by dholmes-ora** [comment]
@jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. You haven't even bothered to address the comments I made on the actual changes.

File: make/autoconf/spec.gmk.in
Comment (dholmes-ora) [reply make/autoconf/spec.gmk.in:459]: To be clear there was a reason that `ASAN_ENABLED` was originally added to spec.gmk.in (though there was a suggestion it could/would be removed later), and no reason has been given for `UBSAN_ENABLED` being needed.

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [inline src/java.base/share/native/launcher/main.c:40]: Why would this need `JNIEXPORT`? This is not a JNI function.

**Comment by jcking** [comment]
> @jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. 

Added reviewers to be 2. I had assumed the bot would take care of getting reviews from both. Apparently it does not, will keep that in mind going forward.

> You haven't even bothered to address the comments I made on the actual changes.

I addressed the comments after integrate, my bad.

File: src/java.base/share/native/launcher/main.c
Comment (jcking) [reply src/java.base/share/native/launcher/main.c:40]: Ugh, apparently resolving doesn't add the comment. I thought it did...

So answering, JNIEXPORT is needed so that the symbol is exported and the linker doesn't remove it. It adds __attribute__((visibility("default")).

**Comment by jcking** [comment]
> @jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. You haven't even bothered to address the comments I made on the actual changes.

Okay, I am dumb. I clicked resolve conversation and assumed it would auto add my comment. Apparently it just drops it and resolves it without saying anything. Sorry!

File: src/java.base/share/native/launcher/main.c
Comment (jcking) [reply src/java.base/share/native/launcher/main.c:37]: Yeah, it is unfortunate. However there is no other way to actually set the defaults nicely. The other alternative is to use environment variables, but they are easy to forget when invoking the launcher manually.

File: make/autoconf/spec.gmk.in
Comment (jcking) [reply make/autoconf/spec.gmk.in:459]: Fair. I'll remove it. My previous comment, that I dropped on the floor. was asking about that basically.

**Comment by dholmes-ora** [comment]
> @dholmes-ora Also etiquette-wise, is it preferred that the commenter resolve the conversation or the author, and the commenter re-open if they feel it is not resolved? I am used to latter workflow, but OpenJDK might have different expectations and I'd like to follow them.

I'm not aware of any specific guidelines. Many people don't even bother with resolving conversations. To me it depends on the nature of the comments as to who should decide if the matter is resolved or not.  It is also unclear of commenting on a "resolved" conversation actually unresolves it again. I don't see an "unresolve" button.

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [reply src/java.base/share/native/launcher/main.c:37]: Does the env-var override the src declaration, or the other way around?

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [inline src/java.base/share/native/launcher/main.c:38]: Please explain why we need to override the default options.

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [reply src/java.base/share/native/launcher/main.c:40]: Okay. The limited documentation on these `___xxx_default_options` functions didn't show the need for any export/visibility attributes.

File: src/java.base/share/native/launcher/main.c
Comment (jcking) [reply src/java.base/share/native/launcher/main.c:38]: Done. Also added a comment regarding the environment variable UBSAN_OPTIONS.

File: src/java.base/share/native/launcher/main.c
Comment (jcking) [reply src/java.base/share/native/launcher/main.c:37]: UBSAN_OPTIONS overrides the src declaration. I added to the comment mentioning that.

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [inline src/java.base/share/native/launcher/main.c:41]: If this is primarily for tests then can't we set the env-var in the test Makefile?

Comment (jcking) [reply src/java.base/share/native/launcher/main.c:41]: Primarily, but its not a requirement. We should also be able to invoke `java` as is. The environment variables should be used to force specific behavior for a single invocation. Otherwise, if one forgets to pass the environment variables, things may not go as expected.

File: src/java.base/share/native/launcher/main.c
Comment (robehn) [inline src/java.base/share/native/launcher/main.c:49]: As I said we have more launcher than 'java', if you put this method here you must put it in all launchers.
I.e. all binaries the call JNI_CreateJavaVM, such out tests, such as jni/daemonDestroy/TestDaemonDestroy.java

Comment (jcking) [reply src/java.base/share/native/launcher/main.c]: Attempting an approach that automatically includes `__ubsan_default_options` in binaries by "automagically" including a source file for anything using `SetupNativeCompilation` with `TYPE` being `EXECUTABLE`.

**Comment by jcking** [comment]
Moving back to draft while I figure out some autoconf magic.

File: src/java.base/share/native/launcher/main.c
Comment (jcking) [reply src/java.base/share/native/launcher/main.c]: Okay, I was able to get the autoconf magic working. Instead of having to copy-paste `__ubsan_default_options` to every launcher, it is instead auto-inserted in `SetupNativeCompilation` by including a source file which exports it for anything that is `EXECUTABLE`.

File: src/java.base/share/native/launcher/main.c
Comment (dholmes-ora) [reply src/java.base/share/native/launcher/main.c]: It really comes down to who we expect to do these ubsan builds and how we expect them to be used. If this is something explicit then using the env-var seems fine to me. I'm not sure why it needs to be enabled by default through this source modification.

I'd like to hear other opinions on this.

Comment (jcking) [reply src/java.base/share/native/launcher/main.c]: Sure. It should be noted that whatever we do here for default options will likely have to be repeated for LSan and ASan eventually. So it should be repeatable and easy to maintain. IMO we should enforce strict defaults and then let manual invocations make it less strict via environment variables.

**Review by dholmes-ora — COMMENTED** [review]
I like this approach better in principle. NOt sure about the location of the file - the build folk will no doubt have more to say on that. Can we not just have a single file and use:
```
#ifdef __cplusplus
extern "C" {
#endif
```
?
Thanks.

**Comment by jcking** [comment]
I tried to use a single file, but the build logic attempts to compile as either C or C++ based on file extensions, and has logic based on it. So if I use `.cpp` and the target is all `.c` odd things happen. The same for the inverse.

**Comment by erikj79** [comment]
> I tried to use a single file, but the build logic attempts to compile as either C or C++ based on file extensions, and has logic based on it. So if I use `.cpp` and the target is all `.c` odd things happen. The same for the inverse.

What happens if you just always use a `.c` file for this? Won't the linker accept that?

**Comment by jcking** [comment]
Nope. Some targets end up passing C++ flags to the C compiler, causing a
failure.

**Comment by theRealAph** [comment]
I guess the advantage to putting this in the build machinery (as opposed to using
`--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?

**Comment by jcking** [comment]
Correct. We will be able to turn it on by default and enforce it, if w so
choose, for everything.

**Comment by robehn** [comment]
> I guess the advantage to putting this in the build machinery (as opposed to using `--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?

It will take a while, look a bit on align issue, we have so much code which go from pointer to large -> small -> large, e.g.
```
    address addr = data() + offset;
    return (ImmutableOopMap*) addr;
```
In this case data() needs to return something with the same alignment as a ptr and offset must even in ptr steps.

**Comment by erikj79** [comment]
> Nope. Some targets end up passing C++ flags to the C compiler, causing a failure.

Ah right, we (mis)use CFLAGS (instead of CXXFLAGS) in some SetupNativeCompilation calls when all source files are C++. In that case, your suggested patch makes sense to me.

**Comment by jcking** [comment]
> > I guess the advantage to putting this in the build machinery (as opposed to using `--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?
> 
> It will take a while, look a bit on align issue, we have so much code which go from pointer to large -> small -> large, e.g.
> 
> ```
>     address addr = data() + offset;
>     return (ImmutableOopMap*) addr;
> ```
> 
> In this case data() needs to return something with the same alignment as a ptr and offset must be in even in ptr steps.

Yeah, some of the cases we may just have to suppress if they are not feasible to fix without lots of effort. My intention is to fix all the low hanging fruit, and then suppress the remaining cases. For the suppressed cases, I'll file P4 bugs for each one. Then if somebody feels like fixing them, they can do it.

**Comment by magicus** [comment]
Sorry for the late review.

"Magic" solutions like this is always a bit scary. But we have some precedence in terms of the "autoheaders" for assembly files, so I guess this is not worse.

However, I do think the included source files should be treated like the autoheaders, and reside in data rather than in `src`. The latter is intended for buildtools, even though they are a bit scattered at the moment (there is a long-standing JBS bug about fixing that), and these are not really intended to be compiled into a product by itself, rather as something "injected" into all builds. I'd recommend `make/data/ubsan`, and perhaps skip the leading `__`.

File: make/autoconf/jdk-options.m4
Comment (magicus) [reply make/autoconf/jdk-options.m4:450]: Well, this function actually does two things -- first it checks if ubsan support should be enabled, and that code really belong here. But then if it i enabled, it adds additional compiler options, including globally silencing some warnings, and that should really reside in flags-cflags.m4. 

But, I see we have this pattern for other optional tooling support like ASan and GCov, so I guess this is fine for now.

I'm really looking forward to a better way of handling flags so we wouldn't have to do it like this.

**Comment by magicus** [comment]
I much also check: is this really needed for *all* executables we make? I would have guessed that it would suffice with the "launchers", i.e. like `java`, `javac`, `jar` etc. These are all compiled from a single source file, `src/java.base/share/native/launcher/main.c`, with different defines depending on what Java classes it should actually launch, and with different output names.

Doing things this way will also affect non-launcher executables, like `jabswitch`, `msiwrapper` and all executable test binaries. 

That might be correct, but I could not be certain of that from trying to read the backlog of discussion in this bug.

**Comment by jcking** [comment]
> However, I do think the included source files should be treated like the autoheaders, and reside in data rather than in `src`. The latter is intended for buildtools, even though they are a bit scattered at the moment (there is a long-standing JBS bug about fixing that), and these are not really intended to be compiled into a product by itself, rather as something "injected" into all builds. I'd recommend `make/data/ubsan`, and perhaps skip the leading `__`.

Moved to the directory suggested.

**Comment by jcking** [comment]
> I much also check: is this really needed for _all_ executables we make? I would have guessed that it would suffice with the "launchers", i.e. like `java`, `javac`, `jar` etc. These are all compiled from a single source file, `src/java.base/share/native/launcher/main.c`, with different defines depending on what Java classes it should actually launch, and with different output names.
> 
> Doing things this way will also affect non-launcher executables, like `jabswitch`, `msiwrapper` and all executable test binaries.
> 
> That might be correct, but I could not be certain of that from trying to read the backlog of discussion in this bug.

For UBSan, the inclusion in every executable is intentional. For upcoming LSan/ASan stuff, it will be just for launchers. The reason is that UBSan is relaxed by default and won't terminate upon undefined behavior. LSan/ASan are strict by default, and for LSan to work we need to disable some things only when the JVM is being used. So for this commit all executables is correct. For ASan/LSan as you suggest only launchers will have them in that case.

**Comment by jcking** [comment]
Friendly poke.

**Comment by robehn** [comment]
I think @magicus is back next week. Since this is mainly build stuff I'll sit this one out.

**Comment by jcking** [comment]
Ah okay, no problem. Thanks for the update.

**Review by magicus — APPROVED** [review]

**Comment by magicus** [comment]
Hi, back from vacation today. Thanks for waiting on my input! This looks good.

**Pull request closed without merging**
